File: symmetric-core/src/main/java/org/jumpmind/symmetric/transport/file/FileOutgoingTransport.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.symmetric.transport.file;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.OutputStream;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;

import org.apache.commons.io.IOUtils;
import org.jumpmind.symmetric.model.ChannelMap;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.model.OutgoingBatch;
import org.jumpmind.symmetric.model.OutgoingBatches;
import org.jumpmind.symmetric.service.IConfigurationService;
import org.jumpmind.symmetric.transport.BatchBufferedWriter;
import org.jumpmind.symmetric.transport.IOutgoingWithResponseTransport;
import org.jumpmind.symmetric.web.WebConstants;

public class FileOutgoingTransport implements IOutgoingWithResponseTransport {

    String fileName;

    BatchBufferedWriter writer;
    
    OutputStream out;

    boolean open = true;
    
    Node remoteNode;
    
    String outgoingDir;
    
    private List<OutgoingBatch> processedBatches;
    
    private String extension;
    
    public FileOutgoingTransport(Node remoteNode, Node localNode, String outgoingDir) throws IOException {
        this.outgoingDir = outgoingDir;
        this.fileName = outgoingDir + File.separator + localNode.getNodeGroupId() + "-" + localNode.getNodeId() + "_to_" + 
                remoteNode.getNodeGroupId() + "-" + remoteNode.getNodeId() + "_" + System.currentTimeMillis();
        this.remoteNode = remoteNode;
    }
    
    public String getOutgoingDir() {
        return outgoingDir;
    }

    @Override
    public BufferedWriter openWriter() {
        try {
            writer = new BatchBufferedWriter(new FileWriter(fileName + ".tmp"));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return writer;
    }
    
    @Override
    public BufferedWriter getWriter() {
        return writer;
    }

    @Override
    public OutputStream openStream() {
        try {
            out = new FileOutputStream(fileName + ".tmp");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return out;
    }

    @Override
    public BufferedReader readResponse() throws IOException {
        List<Long> batchIds = new ArrayList<Long>();
        if (processedBatches != null) {
            for (OutgoingBatch processedBatch : processedBatches) {
                batchIds.add(processedBatch.getBatchId());
            }
        } else if (writer != null) {
            batchIds = writer.getBatchIds();
        }
        
        StringBuilder resp = new StringBuilder();
        for (Long batchId : batchIds) {
            resp.append(WebConstants.ACK_BATCH_NAME).append(batchId).append("=").append(WebConstants.ACK_BATCH_OK).append("&");
            resp.append(WebConstants.ACK_NODE_ID).append(batchId).append("=").append(remoteNode.getNodeId()).append("&");
        }
        return new BufferedReader(new StringReader(resp.toString()));
    }

    @Override
    public void close() {
        IOUtils.closeQuietly(writer);
        IOUtils.closeQuietly(out);
        open = false;
    }

    @Override
    public boolean isOpen() {
        return open;
    }

    @Override
    public ChannelMap getSuspendIgnoreChannelLists(IConfigurationService configurationService, String queue, Node targetNode) {
        return configurationService.getSuspendIgnoreChannelLists();
    }

    public void complete(boolean success) {
        if (!success) {
            new File(fileName + ".tmp").delete();
        } else if (writer != null) {
            new File(fileName + ".tmp").renameTo(new File(fileName + ".csv"));
        } else {
            new File(fileName + ".tmp").renameTo(new File(fileName + ".zip"));
        }
    }

    public List<OutgoingBatch> getProcessedBatches() {
        return processedBatches;
    }

    public void setProcessedBatches(List<OutgoingBatch> processedBatches) {
        this.processedBatches = processedBatches;
    }
}

File: symmetric-client-clib/src/transport/file/FileOutgoingTransport.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "transport/file/FileOutgoingTransport.h"

char *SymFileOutgoingTransport_getFileName(SymFileOutgoingTransport *this) {
    long currentTimeMillis;
    time(&currentTimeMillis);
    currentTimeMillis *= 1000;

    return SymStringUtils_format("%s/%s-%s_to_%s-%s_%ld", this->offlineOutgoingDir,
            this->localNode->nodeGroupId, this->localNode->nodeId,  this->remoteNode->nodeGroupId, this->remoteNode->nodeId, currentTimeMillis);
}

long SymFileOutgoingTransport_process(SymFileOutgoingTransport *this, SymDataProcessor *processor) {

    SymFileUtils_mkdir(this->offlineOutgoingDir);

    processor->open(processor);

    int BUFFER_SIZE = 2048;
    char inputBuffer[BUFFER_SIZE];
    char* fileNameBase = SymFileOutgoingTransport_getFileName(this);
    char *tmpFileName = SymStringUtils_format("%s%s", fileNameBase, ".tmp");
    char *csvFileName = SymStringUtils_format("%s%s", fileNameBase, ".csv");

    SymLog_debug("Writing file %s", tmpFileName);

    long result = SYM_TRANSPORT_SC_SERVICE_UNAVAILABLE;

    FILE *file = fopen(tmpFileName, "w");

    if (file) {
        int size;
        while ((size = processor->process(processor, inputBuffer, 1, BUFFER_SIZE)) > 0) {
            int result = fprintf(file, "%.*s", size, inputBuffer);
            if (result < 0) {
                SymLog_warn("failed to write to file %s rc=%d", tmpFileName, result);
            }
        }
        fclose(file);
        result = SYM_TRANSPORT_OK;
    } else {
        SymLog_error("Failed to open file for writing. %s", fileNameBase);
        result = SYM_TRANSPORT_SC_SERVICE_UNAVAILABLE;
    }

    SymLog_debug("Rename '%s' to '%s'", tmpFileName, csvFileName);
    int renameResult = rename(tmpFileName, csvFileName);
    if (renameResult != 0) {
        SymLog_warn("Failed to rename '%s' to '%s' %s", tmpFileName, csvFileName, strerror(errno));
    }

    free(csvFileName);
    free(tmpFileName);
    free(fileNameBase);

    SymList *batchIds = processor->getBatchesProcessed(processor);
    SymStringBuilder *buff = SymStringBuilder_new(NULL);
    int i;
    for (i = 0; i < batchIds->size; ++i) {
        char* batchId = batchIds->get(batchIds, i);
        buff->append(buff, SYM_WEB_CONSTANTS_ACK_BATCH_NAME);
        buff->append(buff, batchId);
        buff->append(buff, "=");
        buff->append(buff, SYM_WEB_CONSTANTS_ACK_BATCH_OK);
        buff->append(buff, "&");
    }

    this->super.ackString = buff->destroyAndReturn(buff);

    return result;
}


void SymFileOutgoingTransport_destroy(SymFileOutgoingTransport *this) {
    free(this);
}

SymFileOutgoingTransport * SymFileOutgoingTransport_new(SymFileOutgoingTransport *this, SymNode *remoteNode, SymNode *localNode,
        char *offlineOutgoingDir) {
    if (this == NULL) {
        this = (SymFileOutgoingTransport *) calloc(1, sizeof(SymFileOutgoingTransport));
    }
    SymOutgoingTransport *super = (SymOutgoingTransport *)&this->super;
    this->remoteNode = remoteNode;
    this->localNode = localNode;
    this->offlineOutgoingDir = offlineOutgoingDir;

    super->process = (void *) &SymFileOutgoingTransport_process;
    super->destroy = (void *) &SymFileOutgoingTransport_destroy;
    return this;
}
